axelor-stock/src/main/java/com/axelor/apps/stock/service/StoredProductsServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+    storedProducts.setStoredProduct(product);
+    storedProducts.setTrackingNumber(trackingNumber);
+    storedProducts.setCurrentQty(currentQty);
+    storedProducts.setUnit(product.getUnit());
+    storedProducts.setToStockLocation(toStockLocation);
+    storedProducts.setStoredQty(storedQty);
+    storedProducts.setStockMoveLine(stockMoveLine);
+    storedProducts.setMassStockMove(massStockMove);
+    return storedProducts;
+  }
+
+  @Override
+  @Transactional(rollbackOn = Exception.class)
+  public void createStockMoveAndStockMoveLine(StoredProducts storedProducts)
+      throws AxelorException {
+    BigDecimal storedQty =
+        storedProducts.getStoredQty() != null ? storedProducts.getStoredQty() : null;
+    BigDecimal currentQty =
+        storedProducts.getCurrentQty() != null ? storedProducts.getCurrentQty() : null;
+
+    if (storedQty.compareTo(currentQty) == -1 && storedQty.compareTo(BigDecimal.ZERO) != 0) {
+      Product duplicateProduct = storedProducts.getStoredProduct();
+      TrackingNumber duplicateTrackingNumber = storedProducts.getTrackingNumber();
+      StockLocation duplicateToStockLocation = storedProducts.getToStockLocation();
+      StockMoveLine duplicateStockMoveLine = storedProducts.getStockMoveLine();
+      MassStockMove duplicateMassStockMove = storedProducts.getMassStockMove();
+      storedProducts.setCurrentQty(storedQty);
+      StoredProducts duplicateStoredProduct =
+          createStoredProduct(
+              duplicateProduct,
+              duplicateTrackingNumber,
+              currentQty.subtract(storedQty),
+              duplicateToStockLocation,
+              BigDecimal.ZERO,
+              duplicateStockMoveLine,
+              duplicateMassStockMove);
+      storeProductsRepository.save(duplicateStoredProduct);
+    }
+
+    if (storedProducts.getStoredQty().compareTo(BigDecimal.ZERO) != 0
+        && storedProducts.getToStockLocation() != null
+        && storedProducts.getStockMoveLine() == null) {
+
+      if (storedProducts.getStoredQty().compareTo(storedProducts.getCurrentQty()) == 1
+          && storedProducts.getTrackingNumber() != null) {
+        throw new AxelorException(
+            0,
+            I18n.get(StockExceptionMessage.STORED_QUANTITY_GREATER_THAN_CURRENT_QTY)
+                + " "
+                + storedProducts.getStoredProduct().getFullName()
+                + " "
+                + storedProducts.getTrackingNumber().getTrackingNumberSeq());
+      } else if (storedProducts.getStoredQty().compareTo(storedProducts.getCurrentQty()) == 1
+          && storedProducts.getTrackingNumber() == null) {
+        throw new AxelorException(
+            0,
+            I18n.get(StockExceptionMessage.STORED_QUANTITY_GREATER_THAN_CURRENT_QTY)
+                + " "
+                + storedProducts.getStoredProduct().getFullName());
+      }
+      Address toAddress = storedProducts.getToStockLocation().getAddress();
```

Comment:
Please refactor this 
`Address toAddress = storedProducts.getToStockLocation().getAddress();
      Address fromAddress = storedProducts.getMassStockMove().getCartStockLocation().getAddress();
      Company company = storedProducts.getMassStockMove().getCompany();
      StockLocation toStockLocation = storedProducts.getToStockLocation();
      StockLocation fromStockLocation = storedProducts.getMassStockMove().getCartStockLocation();
      Product product = storedProducts.getStoredProduct();
      TrackingNumber trackingNumber =
          storedProducts.getTrackingNumber() != null ? storedProducts.getTrackingNumber() : null;
      BigDecimal qty = storedProducts.getStoredQty() != null ? storedProducts.getStoredQty() : null;
      Unit unit = storedProducts.getUnit() != null ? storedProducts.getUnit() : null;
      StockMove stockMove =
          stockMoveService.createStockMove(
              fromAddress,
              toAddress,
              company,
              fromStockLocation,
              toStockLocation,
              LocalDate.now(),
              LocalDate.now(),
              null,
              StockMoveRepository.TYPE_INTERNAL);
      StockMoveLine stockMoveLine =
          stockMoveLineService.createStockMoveLine(
              stockMove,
              product,
              trackingNumber,
              qty,
              qty,
              unit,
              null,
              fromStockLocation,
              toStockLocation);
      stockMove.setMassStockMove(storedProducts.getMassStockMove());
      stockMoveService.plan(stockMove);
      stockMoveService.realize(stockMove);
      stockMoveRepository.save(stockMove);
      stockMoveLineRepository.save(stockMoveLine);
      storedProducts.setStockMoveLine(stockMoveLine);
      storeProductsRepository.save(storedProducts);` in a method